From ef4508a83bb8afe1d6033507793d204e1874eab9 Mon Sep 17 00:00:00 2001
From: Xiaoyu Huang <quic_xiahua@quicinc.com>
Date: Wed, 11 Jun 2025 12:19:56 +0800
Subject: [PATCH] gl-renderer: Make YUV format choose EXTERNAL_OES texture
 target

Based on EGL spec, YUV format should choose EXTERNAL_OES as texture target
to make it be correctly rendered by EGL.

This change fixed the wrong-color issue when wayland client using dmabuf
protocol to share NV12/P010 linear buffer to server without providing a
valid modifier.

Upstream-Status: Inappropriate [qti specific]

Change-Id: I19f9b3e5d783c4b6f715a729e4c96ef823d1f305
Signed-off-by: Xiaoyu Huang <quic_xiahua@quicinc.com>
---
 libweston/renderer-gl/gl-renderer.c | 5 +++++
 1 file changed, 5 insertions(+)

diff --git a/libweston/renderer-gl/gl-renderer.c b/libweston/renderer-gl/gl-renderer.c
index 28c18a5..ea74a69 100644
--- a/libweston/renderer-gl/gl-renderer.c
+++ b/libweston/renderer-gl/gl-renderer.c
@@ -3209,6 +3209,11 @@ choose_texture_target(struct gl_renderer *gr,
 	}
 
 	switch (attributes->format & ~DRM_FORMAT_BIG_ENDIAN) {
+	case DRM_FORMAT_NV12:
+	case DRM_FORMAT_NV21:
+	case DRM_FORMAT_YUV420:
+	case DRM_FORMAT_YVU420:
+	case DRM_FORMAT_P010:
 	case DRM_FORMAT_YUYV:
 	case DRM_FORMAT_YVYU:
 	case DRM_FORMAT_UYVY:
